src/neurostatslib/base_class.py

```diff
@@ -0,0 +1,578 @@
+"""## Abstract class for estimators."""
+
+import abc
+import inspect
+import warnings
+from collections import defaultdict
+from typing import Any, Literal, Optional, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+from jax._src.lib import xla_client
+from numpy.typing import ArrayLike, NDArray
+
+from .utils import has_local_device, is_sequence
+
+
+class _Base:
+    """Base class for neurostatslib estimators.
+
+    A base class for estimators with utilities for getting and setting parameters,
+    and for interacting with specific devices like CPU, GPU, and TPU.
+
+    This class provides utilities for:
+    - Getting and setting parameters using introspection.
+    - Sending arrays to target devices (CPU, GPU, TPU).
+
+    Parameters
+    ----------
+    **kwargs : dict
+        Arbitrary keyword arguments.
+
+    Attributes
+    ----------
+    _kwargs_keys : list
+        List of keyword arguments provided during the initialization.
+
+    Notes
+    -----
+    The class provides helper methods mimicking scikit-learn's get_params and set_params.
+    Additionally, it has methods for selecting target devices and sending arrays to them.
+    """
+
+    def __init__(self, **kwargs):
+        self._kwargs_keys = list(kwargs.keys())
+        for key in kwargs:
+            setattr(self, key, kwargs[key])
+
+    def get_params(self, deep=True):
+        """
+        From scikit-learn, get parameters by inspecting init.
+
+        Parameters
+        ----------
+        deep
+
+        Returns
+        -------
+            out:
+                A dictionary containing the parameters. Key is the parameter
+                name, value is the parameter value.
+        """
+        out = dict()
+        for key in self._get_param_names():
+            value = getattr(self, key)
+            if deep and hasattr(value, "get_params") and not isinstance(value, type):
+                deep_items = value.get_params().items()
+                out.update((key + "__" + k, val) for k, val in deep_items)
+            out[key] = value
+        # add kwargs
+        for key in self._kwargs_keys:
+            out[key] = getattr(self, key)
+        return out
+
+    def set_params(self, **params: Any):
```

Comment:
as commented above, it would be the only use of sklearn in all the package (the dependency can be removed and the code won't break), but i am fine with having it as a dependency.